test/powershell/engine/Module/ModuleSpecification.Tests.ps1

```diff
@@ -0,0 +1,257 @@
+
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Take a hashtable and return two copies,
+# one with the key/value pair, the other without
+function PermuteHashtableOnProperty
+{
+    param([hashtable[]]$Hashtable, [string]$Key, [object]$Value)
+
+    foreach ($ht in $Hashtable)
+    {
+        $ht.Clone()
+    }
+
+    foreach ($ht in $Hashtable)
+    {
+        $ht2 = $ht.Clone()
+        $ht2.$Key = $Value
+        $ht2
+    }
+}
+
+# Take a base hashtable and produce all possible
+# combinations of that hashtable with the keys
+# and values in the key/value hashtable
+function PermuteHashtable
```

Comment:
Can the test structures be initialized statically? It seems to be more simple and visual than to understand this helper code. What do you think?

Reply:
I've been thinking about this. You're definitely right that the number of possible ModuleSpecification field combinations is lower than I imagined initially.

On the other hand, for every field that could be added to a ModuleSpecification, we would have to generate exponentially more tests vs add a single field in here. Perhaps that's unlikely. But I can see something like PSEdition being added, which would take the number of static cases from 6 to 12.